### PR TITLE
Scripts/SmartScripts: Fix charmed behavior for SmartAI

### PR DESCRIPTION
**Pull request by Treeston**
# hype

Fix SmartAI to no longer control movement or cast spell while the creature is controlled by a player.
Also add new `SMART_EVENT_FLAG_WHILE_CHARMED` event flag that allows overriding this behavior.

### CODE REVIEW

File: src/server/game/AI/SmartScripts/SmartAI.cpp [hunk 1]
```diff
@@ -545,8 +576,12 @@ void SmartAI::JustReachedHome()
 
 void SmartAI::EnterCombat(Unit* enemy)
 {
-    me->InterruptNonMeleeSpells(false); // must be before ProcessEvents
     GetScript()->ProcessEventsFor(SMART_EVENT_AGGRO, enemy);
+
+    if (!IsAIControlled())
+        return;
+
+    me->InterruptNonMeleeSpells(false); // must be before ProcessEvents
```
Comment (Shauren) [inline src/server/game/AI/SmartScripts/SmartAI.cpp]: well this is obviously not good - see the comment

**Comment by ccrs** [comment]
"TODO: Some sort of system for smart actions to indicate whether they want to be executed during charm or not."

We talked about that months ago :smile:
Add a event flag similar to the ones that already exist to set which ones should trigger while charmed or not. Thats the easy part... the fun part comes when converting the tons of scripts that now are triggered when charmed (and usually they just check auras :D).

**Comment by ccrs** [comment]
Eitherway I'm sure this already breaks a considerable amount of scripts xd

**Comment by Treeston** [comment]
@ccrs the difference is that this time, it's a matter of "everything still _works_ while charmed, now we need to selectively figure out what _shouldn't_ work". Before it was "nothing works while charmed".

**Comment by ccrs** [comment]
Casts are not the only thing that is made while charmed (texts, explored credit, ...)
IMO the param should be on the event (like the param that currrently exists to not repeat the event) instead of the cast.

**Comment by Treeston** [comment]
Yes, but texts etc should all still fire while charmed, no? The only thing we want to prevent are things that "fight" for control with the controller (casts and movement). Movement is prevented (allowing it is pointless, since the controller can always override it), and casting has the flag.

**Comment by ccrs** [comment]
but then again, right now the only way to mark events for "charm only" is with link on aura for example or using conditions. How simple would it be just marking a flag? xd

**Comment by Treeston** [comment]
That's kind of the intended use case for conditions though.

**Comment by ccrs** [comment]
then why using the smart cast flag of combat move? its the same logic

Plus some charms cannot be "detected" with conditions.

**Comment by Treeston** [comment]
What kind of charms? `IsCharmed` should still be true?

PS: If there's no condition for charmed, then we need one of those instead.

**Comment by Treeston** [comment]
Implementing "_only_ while charmed" using SAI flags is kind of silly anyway. If anything, we'd have a flag for "can be done while charmed", otherwise it'd be "only while not charmed".

**Comment by Treeston** [comment]
Re conditions: 9785546

**Comment by ccrs** [comment]
I dont see how making an event flag is "dumber" than have to create one separate condition for every single event that has to trigger only on charm :P

**Comment by Treeston** [comment]
It's gonna cost us some storage space because event flags now exceed `uint8`, but there you go @ccrs

**Comment by ccrs** [comment]
its the only way to not rewrite every script with charm mechanic :)

**Comment by Treeston** [comment]
It's still gonna break some scripts, but those can be fixed manually. Can't make this omelet without breaking some eggs, I'm afraid.

**Comment by Killyana** [comment]
This also fix the issue with vehicles using sai?

**Comment by Treeston** [comment]
Yes.

**Comment by Treeston** [comment]
Merged in 96f9038 e9eacae

**Comment by lost-illusion** [comment]
@Treeston Hi, can i ask you about this commit. For example we have spell 32802, and this spell summon npc 18734. In SummonPropertiesEntry type seted to SUMMON_TYPE_GUARDIAN, and after summon npc "controlled by player" and don't cast OOC Healing player (In update you not set flag for spell casts) . Can you explain me please why ? I mean it guard, player can't directly controll her. Or we should manualy set flag if needed ? I'm sorry if i understand this wrong.

**Comment by Treeston** [comment]
Flag should be manually set for any guardians that should use it, @lost-illusion. Open an issue about your specific case, it's a very quick fix.

**Comment by Treeston** [comment]
0f73fc8

**Comment by Treeston** [comment]
See 0f7efc8

**Comment by lost-illusion** [comment]
@Treeston  Thanks. Maybe we should found all guardians and check their ? I mean non controlled by player. Minions, guardians.

**Comment by Treeston** [comment]
If you find any guardians missing the flag, open an issue for it.

**Comment by lost-illusion** [comment]
Ok.

**Pull request closed without merging**